test/t/conftest.py

```diff
@@ -186,29 +186,39 @@ def partialize(
 def bash(request) -> pexpect.spawn:
 
     logfile = None
-    if os.environ.get("BASHCOMP_TEST_LOGFILE"):
-        logfile = open(os.environ["BASHCOMP_TEST_LOGFILE"], "w")
-    elif os.environ.get("CI"):
-        logfile = sys.stdout
-    testdir = os.path.abspath(
-        os.path.join(os.path.dirname(__file__), os.pardir)
-    )
-    env = os.environ.copy()
-    env.update(
-        dict(
-            SRCDIR=testdir,  # TODO needed at least by bashrc
-            SRCDIRABS=testdir,
-            PS1=PS1,
-            INPUTRC="%s/config/inputrc" % testdir,
-            TERM="dumb",
-            LC_COLLATE="C",  # to match Python's default locale unaware sort
-            HISTFILE="/dev/null",  # to leave user's history file alone
-        )
-    )
-
+    histfile = None
     tmpdir = None
     bash = None
     try:
+
+        if os.environ.get("BASHCOMP_TEST_LOGFILE"):
+            logfile = open(os.environ["BASHCOMP_TEST_LOGFILE"], "w")
+        elif os.environ.get("CI"):
+            logfile = sys.stdout
+
+        testdir = os.path.abspath(
+            os.path.join(os.path.dirname(__file__), os.pardir)
+        )
+
+        # create temporary file for HISTFILE
```

Comment:
Let's add a comment here and modify the commit message to elaborate why we're doing the tempfile here -- I think it is no longer obvious now that we're also unsetting `HISTFILE`. I know that at least I would have likely forgotten about the details in a year or two and would think of it as an oversight ripe fore cleanup :)

Reply:
OK, thank you for your comments. I fixed up the last commit with the following change. I also added a description in the commit log. ce3583f6

<details><summary><code>git diff -b 60c5543 ce3583f</code></summary>

```diff
$ git diff -b 60c5543 ce3583f
diff --git a/test/t/conftest.py b/test/t/conftest.py
index fab52d6b..29d7418b 100644
--- a/test/t/conftest.py
+++ b/test/t/conftest.py
@@ -189,7 +189,6 @@ def bash(request) -> pexpect.spawn:
     histfile = None
     tmpdir = None
     bash = None
-    try:

     if os.environ.get("BASHCOMP_TEST_LOGFILE"):
         logfile = open(os.environ["BASHCOMP_TEST_LOGFILE"], "w")
@@ -200,10 +199,20 @@ def bash(request) -> pexpect.spawn:
         os.path.join(os.path.dirname(__file__), os.pardir)
     )

-        # create temporary file for HISTFILE
+    # Create an empty temporary file for HISTFILE.
+    #
+    # To prevent the tested Bash processes from writing to the user's
+    # history file or any other files, we will unset the variable
+    # "HISTFILE" in "test/config/bashrc".  However, the history is
+    # loaded before the first prompt is shown and "test/config/bashrc"
+    # is loaded.  We create an empty history file using tempfile for a
+    # transient value of HISTFILE.
     histfile = tempfile.NamedTemporaryFile(
         prefix="bash-completion-test_", delete=False
     )
+
+    try:
+        # release the file handle so that Bash can open the file.
         histfile.close()

         env = os.environ.copy()
@@ -317,7 +326,10 @@ def bash(request) -> pexpect.spawn:
         if tmpdir:
             tmpdir.cleanup()
         if histfile:
+            try:
                 os.remove(histfile.name)
+            except OSError:
+                pass
         if logfile and logfile != sys.stdout:
             logfile.close()

```
</details>

1. I moved the position of `try:` to immediately before `histfile.close()`.
2. I added `try:`...`except OSError: pass` in case we failed to remove `histfile` (possibly after failing `histfile.close()` as you commented) in the cleanup code.
3. I added comment for the reasoning for the temporary file.

But maybe we don't have to unset `HISTFILE` anymore because we now have a temporary file which can be used as the output target of the command history. Just using the temporary file as `HISTFILE` is clearer than combining it with `unset -v HISTFILE` and thus easy to explain. Nevertheless we may explain why we didn't choose to use neither of `HISTFILE=/dev/null` or `unset -v HISTILFE`

Reply:
Agreed, clearer to remove the `unset` now that we have the tempfile.

Reply:
May I confirm what is that "👍" for? Can I remove `unset -v HISTFILE` from `test/config/bashrc`?

I actually noticed that `test/config/bashrc` is linked from a symbolic link `docs/bashrc`, so I feel we shouldn't include test-specific codes in `test/config/bashrc`. Or, if we keep the `unset -v HISTFILE` line in `test/config/bashrc`, we may create a real copy at `docs/bashrc` that doesn't have `unset -v HISTFILE`.

Reply:
I have updated the commit. 0ee13b2